framework/src/meshgenerators/MeshDiagnosticsGenerator.C

```diff
@@ -0,0 +1,1196 @@
+      // average as the group of unrefined elements we will call this good enough for now
+      // For tetrahedral elements we cannot rely on the children all matching as the choice in
+      // the diagonal selection can be made differently. We'll just say 4 matching children is
+      // good enough for the heuristic
+      unsigned int num_children_match = 0;
+      for (const auto & child : parent_ptr->child_ref_range())
+      {
+        for (const auto & potential_children : fine_elements)
+          if (MooseUtils::absoluteFuzzyEqual(child.vertex_average()(0),
+                                             potential_children->vertex_average()(0),
+                                             _non_conformality_tol) &&
+              MooseUtils::absoluteFuzzyEqual(child.vertex_average()(1),
+                                             potential_children->vertex_average()(1),
+                                             _non_conformality_tol) &&
+              MooseUtils::absoluteFuzzyEqual(child.vertex_average()(2),
+                                             potential_children->vertex_average()(2),
+                                             _non_conformality_tol))
+          {
+            num_children_match++;
+            break;
+          }
+      }
+
+      if (num_children_match == num_children ||
+          ((elem_type == TET4 || elem_type == TET10 || elem_type == TET14) &&
+           num_children_match == 4))
+      {
+        num_likely_AMR_created_nonconformality++;
+        if (num_likely_AMR_created_nonconformality < 10)
+        {
+          _console << "Detected non-conformality likely created by AMR near" << *node
+                   << Moose::stringify(elem_type)
+                   << " elements that could be merged into a coarse element:" << std::endl;
+          for (const auto & elem : fine_elements)
+            _console << elem->id() << " ";
+          _console << std::endl;
+        }
+        else if (num_likely_AMR_created_nonconformality == 10)
+          _console << "Maximum log output reached, silencing output" << std::endl;
+      }
+    }
+
+    diagnosticsLog(
+        "Number of non-conformal nodes likely due to mesh refinement detected by heuristic: " +
+            Moose::stringify(num_likely_AMR_created_nonconformality),
+        _check_adaptivity_non_conformality,
+        num_likely_AMR_created_nonconformality);
+    pl->unset_close_to_point_tol();
+  }
+
+  if (_check_local_jacobian != "NO_CHECK")
+  {
+    unsigned int num_negative_elem_qp_jacobians = 0;
+    // Check elements (assumes serialized mesh)
+    for (const auto & elem : mesh->element_ptr_range())
+    {
+      // Get a high-ish order quadrature
+      QGauss qrule(mesh->mesh_dimension(), FOURTH);
+
+      // Use the default element order
+      // We would error if going above what the element can't support
+      const FEType fe_type(elem->default_order(), libMesh::LAGRANGE);
```

Comment:
I'd just create a constant `MONOMIAL` here.  The map is the map regardless, and if we're going to have to have shape functions because that's the easiest API to use, we should at least evaluate the cheapest shape functions we can ... and for that matter, we shouldn't evaluate them - prerequest `get_JxW` to tell the `FE` that it needs to evaluate the map, and then it won't fall back on the (deprecated!) "I got no prerequests, best evaluate everything to be safe" code path.

Reply:
changed